File: dkplayer-ui/src/main/java/com/dueeeke/videocontroller/component/VodControlView.java
package com.dueeeke.videocontroller.component;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.Animation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.dueeeke.videocontroller.R;
import com.dueeeke.videoplayer.controller.IControlComponent;
import com.dueeeke.videoplayer.controller.MediaPlayerControlWrapper;
import com.dueeeke.videoplayer.player.VideoView;
import com.dueeeke.videoplayer.util.PlayerUtils;

import static com.dueeeke.videoplayer.util.PlayerUtils.stringForTime;

/**
 * 点播底部控制栏
 */
public class VodControlView extends FrameLayout implements IControlComponent, View.OnClickListener, SeekBar.OnSeekBarChangeListener {
    
    protected MediaPlayerControlWrapper mMediaPlayerWrapper;

    private TextView mTotalTime, mCurrTime;
    private ImageView mFullScreen;
    private LinearLayout mBottomContainer;
    private SeekBar mVideoProgress;
    private ProgressBar mBottomProgress;
    private ImageView mPlayButton;

    private boolean mIsDragging;

    private boolean mIsShowBottomProgress = true;

    public VodControlView(@NonNull Context context) {
        super(context);
    }

    public VodControlView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public VodControlView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }
    
    
    {
        setVisibility(GONE);
        LayoutInflater.from(getContext()).inflate(getLayoutId(), this, true);
        mFullScreen = findViewById(R.id.fullscreen);
        mFullScreen.setOnClickListener(this);
        mBottomContainer = findViewById(R.id.bottom_container);
        mVideoProgress = findViewById(R.id.seekBar);
        mVideoProgress.setOnSeekBarChangeListener(this);
        mTotalTime = findViewById(R.id.total_time);
        mCurrTime = findViewById(R.id.curr_time);
        mPlayButton = findViewById(R.id.iv_play);
        mPlayButton.setOnClickListener(this);
        mBottomProgress = findViewById(R.id.bottom_progress);
    }

    protected int getLayoutId() {
        return R.layout.dkplayer_layout_vod_control_view;
    }

    /**
     * 是否显示底部进度条，默认显示
     */
    public void showBottomProgress(boolean isShow) {
        mIsShowBottomProgress = isShow;
    }

    @Override
    public void attach(@NonNull MediaPlayerControlWrapper mediaPlayerWrapper) {
        mMediaPlayerWrapper = mediaPlayerWrapper;
    }

    @Override
    public View getView() {
        return this;
    }

    @Override
    public void show(Animation showAnim) {
        mBottomContainer.setVisibility(VISIBLE);
        if (showAnim != null) {
            mBottomContainer.startAnimation(showAnim);
        }
        if (mIsShowBottomProgress) {
            mBottomProgress.setVisibility(GONE);
        }
    }

    @Override
    public void hide(Animation hideAnim) {
        mBottomContainer.setVisibility(GONE);
        if (hideAnim != null) {
            mBottomContainer.startAnimation(hideAnim);
        }
        if (mIsShowBottomProgress) {
            mBottomProgress.setVisibility(VISIBLE);
        }
    }

    @Override
    public void onPlayStateChanged(int playState) {
        switch (playState) {
            case VideoView.STATE_IDLE:
            case VideoView.STATE_PLAYBACK_COMPLETED:
                setVisibility(GONE);
                mBottomProgress.setProgress(0);
                mBottomProgress.setSecondaryProgress(0);
                mVideoProgress.setProgress(0);
                mVideoProgress.setSecondaryProgress(0);
                break;
            case VideoView.STATE_START_ABORT:
            case VideoView.STATE_PREPARING:
            case VideoView.STATE_PREPARED:
            case VideoView.STATE_ERROR:
                setVisibility(GONE);
                break;
            case VideoView.STATE_PLAYING:
                mPlayButton.setSelected(mMediaPlayerWrapper.isPlaying());
                if (mIsShowBottomProgress) {
                    if (mMediaPlayerWrapper.isShowing()) {
                        mBottomProgress.setVisibility(GONE);
                        mBottomContainer.setVisibility(VISIBLE);
                    } else {
                        mBottomContainer.setVisibility(GONE);
                        mBottomProgress.setVisibility(VISIBLE);
                    }
                } else {
                    mBottomContainer.setVisibility(GONE);
                }
                setVisibility(VISIBLE);
                //开始刷新进度
                mMediaPlayerWrapper.startProgress();
                break;
            case VideoView.STATE_PAUSED:
            case VideoView.STATE_BUFFERING:
            case VideoView.STATE_BUFFERED:
                mPlayButton.setSelected(mMediaPlayerWrapper.isPlaying());
                break;
        }
    }

    @Override
    public void onPlayerStateChanged(int playerState) {
        switch (playerState) {
            case VideoView.PLAYER_NORMAL:
                mFullScreen.setSelected(false);
                break;
            case VideoView.PLAYER_FULL_SCREEN:
                mFullScreen.setSelected(true);
                break;
        }
    }

    @Override
    public void adjustView(int orientation, int space) {
        if (orientation == ActivityInfo.SCREEN_ORIENTATION_PORTRAIT) {
            mBottomContainer.setPadding(0, 0, 0, 0);
            mBottomProgress.setPadding(0, 0, 0, 0);
        } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE) {
            mBottomContainer.setPadding(space, 0, 0, 0);
            mBottomProgress.setPadding(space, 0, 0, 0);
        } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE) {
            mBottomContainer.setPadding(0, 0, space, 0);
            mBottomProgress.setPadding(0, 0, space, 0);
        }
    }

    @Override
    public void setProgress(int duration, int position) {
        if (mIsDragging) {
            return;
        }

        if (mVideoProgress != null) {
            if (duration > 0) {
                mVideoProgress.setEnabled(true);
                int pos = (int) (position * 1.0 / duration * mVideoProgress.getMax());
                mVideoProgress.setProgress(pos);
                mBottomProgress.setProgress(pos);
            } else {
                mVideoProgress.setEnabled(false);
            }
            int percent = mMediaPlayerWrapper.getBufferedPercentage();
            if (percent >= 95) { //解决缓冲进度不能100%问题
                mVideoProgress.setSecondaryProgress(mVideoProgress.getMax());
                mBottomProgress.setSecondaryProgress(mBottomProgress.getMax());
            } else {
                mVideoProgress.setSecondaryProgress(percent * 10);
                mBottomProgress.setSecondaryProgress(percent * 10);
            }
        }

        if (mTotalTime != null)
            mTotalTime.setText(stringForTime(duration));
        if (mCurrTime != null)
            mCurrTime.setText(stringForTime(position));
    }

    @Override
    public void onLock() {
        hide(null);
    }

    @Override
    public void onUnlock() {
        show(null);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.fullscreen) {
            toggleFullScreen();
        } else if (id == R.id.iv_play) {
            mMediaPlayerWrapper.togglePlay();
        }
    }

    /**
     * 横竖屏切换
     */
    private void toggleFullScreen() {
        Activity activity = PlayerUtils.scanForActivity(getContext());
        mMediaPlayerWrapper.toggleFullScreen(activity);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        mIsDragging = true;
        mMediaPlayerWrapper.stopProgress();
        mMediaPlayerWrapper.stopFadeOut();
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        long duration = mMediaPlayerWrapper.getDuration();
        long newPosition = (duration * seekBar.getProgress()) / mVideoProgress.getMax();
        mMediaPlayerWrapper.seekTo((int) newPosition);
        mIsDragging = false;
        mMediaPlayerWrapper.startProgress();
        mMediaPlayerWrapper.startFadeOut();
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (!fromUser) {
            return;
        }

        long duration = mMediaPlayerWrapper.getDuration();
        long newPosition = (duration * progress) / mVideoProgress.getMax();
        if (mCurrTime != null)
            mCurrTime.setText(stringForTime((int) newPosition));
    }
}

File: dkplayer-ui/src/main/java/xyz/doikki/videocontroller/component/VodControlView.kt
package xyz.doikki.videocontroller.component

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ActivityInfo
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.annotation.LayoutRes
import xyz.doikki.videocontroller.R
import xyz.doikki.videoplayer.VideoView
import xyz.doikki.videoplayer.util.PlayerUtils
import xyz.doikki.videoplayer.util.orDefault

/**
 * 点播底部控制栏
 */
open class VodControlView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : BaseControlComponent(context, attrs, defStyleAttr),
    View.OnClickListener,
    OnSeekBarChangeListener {

    private var bottomContainer: LinearLayout? = null
    private var fullScreen: ImageView? = null
    private var totalTime: TextView? = null
    private var currTime: TextView? = null
    private var playButton: ImageView? = null
    private var videoProgress: SeekBar? = null
    private var bottomProgress: ProgressBar? = null

    /**
     * 是否正在拖动SeekBar
     */
    private var trackingTouch = false

    /**
     * 是否显示底部进度条，默认显示
     */
    var showBottomProgress = true

    @LayoutRes
    protected open val layoutId: Int = R.layout.dkplayer_layout_vod_control_view


    override fun onVisibilityChanged(isVisible: Boolean, anim: Animation?) {
        if (isVisible) {
            bottomContainer?.let { bottomContainer ->
                bottomContainer.visibility = VISIBLE
                anim?.let {
                    bottomContainer.startAnimation(it)
                }
            }
            if (showBottomProgress) {
                bottomProgress?.visibility = GONE
            }
        } else {
            bottomContainer?.let { bottomContainer ->
                bottomContainer.visibility = GONE
                anim?.let {
                    bottomContainer.startAnimation(it)
                }
            }

            if (showBottomProgress) {
                bottomProgress?.let { bottomProgress ->
                    bottomProgress.visibility = VISIBLE
                    val animation = AlphaAnimation(0f, 1f)
                    animation.duration = 300
                    bottomProgress.startAnimation(animation)
                }
            }
        }
    }

    override fun onPlayStateChanged(playState: Int, extras: HashMap<String, Any>) {
        when (playState) {
            VideoView.STATE_IDLE, VideoView.STATE_PLAYBACK_COMPLETED -> {
                visibility = GONE
                bottomProgress?.let {
                    it.progress = 0
                    it.secondaryProgress = 0
                }
                videoProgress?.let {
                    it.progress = 0
                    it.secondaryProgress = 0
                }
            }
            VideoView.STATE_START_ABORT, VideoView.STATE_PREPARING,
            VideoView.STATE_PREPARED, VideoView.STATE_ERROR -> visibility = GONE
            VideoView.STATE_PLAYING -> {
                playButton?.isSelected = true
                if (showBottomProgress) {
                    if (controller?.isShowing.orDefault()) {
                        bottomProgress?.visibility = GONE
                        bottomContainer?.visibility = VISIBLE
                    } else {
                        bottomContainer?.visibility = GONE
                        bottomProgress?.visibility = VISIBLE
                    }
                } else {
                    bottomContainer?.visibility = GONE
                }
                visibility = VISIBLE
                //开始刷新进度
                controller?.startUpdateProgress()
            }
            VideoView.STATE_PAUSED -> playButton?.isSelected = false
            VideoView.STATE_BUFFERING -> {
                playButton?.isSelected = player?.isPlaying.orDefault()
                // 停止刷新进度
                controller?.stopUpdateProgress()
            }
            VideoView.STATE_BUFFERED -> {
                playButton?.isSelected = player?.isPlaying.orDefault()
                //开始刷新进度
                controller?.startUpdateProgress()
            }
        }
    }

    @SuppressLint("SwitchIntDef")
    override fun onScreenModeChanged(screenMode: Int) {
        when (screenMode) {
            VideoView.SCREEN_MODE_NORMAL -> fullScreen?.isSelected = false
            VideoView.SCREEN_MODE_FULL -> fullScreen?.isSelected = true
        }
        val activity = this.activity ?: return
        val controller = controller ?: return
        val bottomContainer = bottomContainer ?: return
        val bottomProgress = bottomProgress ?: return
        if (controller.hasCutout == true) {
            val orientation = activity.requestedOrientation
            val cutoutHeight = controller.cutoutHeight
            when (orientation) {
                ActivityInfo.SCREEN_ORIENTATION_PORTRAIT -> {
                    bottomContainer.setPadding(0, 0, 0, 0)
                    bottomProgress.setPadding(0, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE -> {
                    bottomContainer.setPadding(cutoutHeight, 0, 0, 0)
                    bottomProgress.setPadding(cutoutHeight, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE -> {
                    bottomContainer.setPadding(0, 0, cutoutHeight, 0)
                    bottomProgress.setPadding(0, 0, cutoutHeight, 0)
                }
            }
        }
    }


    override fun onLockStateChanged(isLocked: Boolean) {
        onVisibilityChanged(!isLocked, null)
    }

    override fun onClick(v: View) {
        val id = v.id
        if (id == R.id.fullscreen) {
            toggleFullScreen()
        } else if (id == R.id.iv_play) {
            controller?.togglePlay()
        }
    }

    /**
     * 横竖屏切换
     */
    private fun toggleFullScreen() {
        controller?.toggleFullScreen()
        // 下面方法会根据适配宽高决定是否旋转屏幕
//        mControlWrapper.toggleFullScreenByVideoSize(activity);
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
        trackingTouch = true
        controller?.let {
            it.stopUpdateProgress()
            it.stopFadeOut()
        }
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        try {
            controller?.let { controller ->
                val player = this.player ?: return@let
                val duration = player.duration
                val newPosition = duration * seekBar.progress / videoProgress!!.max
                player.seekTo(newPosition.toInt().toLong())
                trackingTouch = false
                controller.startUpdateProgress()
                controller.startFadeOut()
            }
        } finally {
            trackingTouch = false
        }
    }


    override fun onProgressChanged(duration: Int, position: Int) {
        if (trackingTouch) {
            return
        }
        videoProgress?.let { seekBar ->
            if (duration > 0) {
                seekBar.isEnabled = true
                val pos = (position * 1.0 / duration * seekBar.max).toInt()
                seekBar.progress = pos
                bottomProgress?.progress = pos
            } else {
                seekBar.isEnabled = false
            }
            val percent = player?.bufferedPercentage.orDefault()
            if (percent >= 95) { //解决缓冲进度不能100%问题
                seekBar.secondaryProgress = seekBar.max
                bottomProgress?.secondaryProgress = bottomProgress?.max.orDefault(100)
            } else {
                seekBar.secondaryProgress = percent * 10
                bottomProgress?.secondaryProgress = percent * 10
            }
        }

        totalTime?.text = PlayerUtils.stringForTime(duration)
        currTime?.text = PlayerUtils.stringForTime(position)
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        if (!fromUser) {
            return
        }
        controller?.playerControl?.let { player ->
            val duration = player.duration
            val newPosition = duration * progress / videoProgress?.max.orDefault(100)
            currTime?.text = PlayerUtils.stringForTime(newPosition.toInt())
        }
    }

    init {
        visibility = GONE
        layoutInflater.inflate(layoutId, this)
        fullScreen = findViewById(R.id.fullscreen)
        fullScreen?.setOnClickListener(this)
        bottomContainer = findViewById(R.id.bottom_container)
        videoProgress = findViewById(R.id.seekBar)
        videoProgress?.setOnSeekBarChangeListener(this)
        //5.1以下系统SeekBar高度需要设置成WRAP_CONTENT
        if (Build.VERSION.SDK_INT <= 22) {
            videoProgress?.layoutParams?.height = ViewGroup.LayoutParams.WRAP_CONTENT
        }
        totalTime = findViewById(R.id.total_time)
        currTime = findViewById(R.id.curr_time)
        playButton = findViewById(R.id.iv_play)
        playButton?.setOnClickListener(this)
        bottomProgress = findViewById(R.id.bottom_progress)
    }
}